### ISSUE TITLE
Duplicate code in VectorList specializations

### ISSUE THREAD

**Issue by davidhesselbom** [magic-lang/ooc-kean#384]
Currently, `FloatVectorList` is almost 400 lines long and consists mostly of vector operation implementations like
```ooc
	operator * (other: This) -> This {
		minimumCount := this count < other count ? this count : other count
		result := This new(minimumCount)
		for (i in 0 .. minimumCount)
			result add(this[i] * other[i])
		result
	}
```
... and then a number of other highly similar functions where the only difference is the operator used on line 5 (in this case, `*`). There has to be a way to reduce the amount of duplicate code - perhaps by using a `map` or `apply` function, if this can be done efficiently.

Several functions also exist in several different specializations, but the code is identical in both/all specializations. Is there really no way to write them in a generic form?

**Comment by marcusnaslund** [comment]
The map() and fold() functions don't really work either.

**Comment by sebastianbaginski** [comment]
And the `sort()` version that takes custom comparator function :) 
Take a look at this https://github.com/cogneco/ooc-kean/pull/327
In general, you can assume anything involving generics to be broken.

**Comment by davidhesselbom** [comment]
Not everything, but maybe the things that work are an exception rather than a rule...

**Comment by marcusnaslund** [comment]
I've been experimenting with this. As it turns out, passing a function which takes two generic types does not work. Passing two generic pointers does not work either. However, passing a function which takes three generic pointers **does work**, for reasons completely unknown. 
For example, I added a zipWith function for VectorList . 
This does not work:
```ooc
zipWith: func <T, S> (S: Class, function: Func(T, T) -> T, other: This<T>) -> This<S> {
	//for every i...
	result := function(this[i], other[i])
}
```
But this does:
```ooc
zipWith: func <T, S> (S: Class, function: Func(S*, T*, T*), other: This<T>) -> This<S> {
	result : S
	//for every i...
	function(value&, this[i]&, other[i]&)
}
```

**Comment by davidhesselbom** [comment]
Does it work for both value and reference types for `S` and `T`?

**Comment by marcusnaslund** [comment]
No, I run into problems when the function has to return something.

**Comment by davidhesselbom** [comment]
When? For value types, or for reference types?

**Comment by marcusnaslund** [comment]
Always. I can't read the correct values inside the function unless I pass by reference, and I can't have the function return anything (so the first argument is a pointer of type S to where the result can be stored).
Then I can implement the + operator in FloatVectorList like this:
```ooc
addFunction := func(result, a, b: Float*) { result@ = a@ + b@ }
this zipWith(Float, addFunction, other) as FloatVectorList

**Comment by marcusnaslund** [comment]
Just confirmed, this type of map() works:
```ooc
map: func ~general (S: Class, function: Func(S*, T*)) -> This<T> {
		result := This new(this count)
		value: S
		for (i in 0 .. this count) {
			function(value&, this[i]&)
			result add(value)
		}
		result
	}
```
Then we can do this (FloatMatrixVector + Float):
```ooc
operator + (value: Float) -> This {
		function := func(result, original: Float*) { result@ = original@ + value }
		this map(Float, function) as FloatVectorList
	}
```
I guess we should also keep the old map,
```ooc
map: func <T, S> (S: Class, function: Func(T) -> S) -> This<S> {
		result := This<S> new(this count)
		for (i in 0 .. this count) {
			result add(function(this[i]))
		}
		result
	}
```
because then we can still do things like:
```ooc
exp: func -> This {
		this map(Float, |value| value exp()) as FloatVectorList
	}
```
The latter approach looks nice but is very limited. Or what do you think? @simonmika ?

**Comment by sebastianbaginski** [comment]
```ooc
operator + (other: This) -> This {
addFunction := func(result, a, b: Float*) { result@ = a@ + b@ }
this zipWith(Float, addFunction, other) as FloatVectorList
}
```
I'm curious about performance here. For a vector of size N this requires N function calls and 3*N pointer dereference operations, compared to simple 3*N memory access block by block in previous version (which we could optimize further using SSE-type of instructions if we like). As far as I'm concerned, rock is not the best compiler when it comes to inline functions :)
In ideal world it should be possible to fix such thing at compile time without introducing any extra penalty at runtime. But here, in the world of ooc, I'd rather have code that is ugly and duplicated but fast :)
Anyway, I would do heavy testing for performance before replacing the existing `operator +` definition with the one above.
I'm just curious, have you tried to run any tests to compare the performance of the two `operator +` implementations ?

**Comment by marcusnaslund** [comment]
Good point, I haven't tested for performance but will look into it.

**Comment by marcusnaslund** [comment]
For 10 000 elements, it takes about 1.9x as long using zipWith.
On average 52 microseconds for the old version and 97 for the new version.

**Comment by simonmika** [comment]
Maybe we should stop using `Closure` which is a cover here but instead start relying on a new class that we might call `Delegate` or something. Maybe then the type issue will go away. Could somebody try?

**Comment by simonmika** [comment]
And I don't like the pointer solution. Is it not better to wrap covers in `Cell`?

**Comment by marcusnaslund** [comment]
Hmm, is there something we can do with the new/upcoming rock version? @thomasfanell

**Comment by simonmika** [comment]
Now that we have owners on closures I would like to test to use more general methods on all `List`s:
- [ ] `map` every element in a list to a iterator of new elements of potentially another type.
- [ ] `apply` function on all element
- [ ] `fold` folds list into new data structure, useful for sums and a lot of other cases
- [ ] `exists` true if at least one element fulfills the predicate
- [ ] `all` true if all elements fulfills the predicate
- [ ] `filter` returns a iterator of all elements fulfilling the predicate

**Comment by sebastianbaginski** [comment]
Do `Closure`s work with covers now ? That'd be great.

**Comment by marcusnaslund** [comment]
This will be fixed well enough by #1208
